### PR TITLE
nuttx/sched: remove nxsched_remove_blocked from up_unblock_task

### PR DESCRIPTION
**Pull request by zyfeier**
## Summary

It takes about 10 cycles to obtain the task list according to the task status. In most cases, we know the task status, so we can directly delete the task from the specified task list to reduce time consuming.

## Impact

Performance:

sem_post -> sem_wait (cycle count):
before:
`171`

after:
`161`

## Testing

sabre-6quad:nsh ostest

### CODE REVIEW

**Comment by masayuki2009** [comment]
```
====================================================================================
Configuration/Tool: stm32f103-minimum/sensors,CONFIG_ARM_TOOLCHAIN_GNU_EABI
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Disabling CONFIG_ARM_TOOLCHAIN_GNU_EABI
  Enabling CONFIG_ARM_TOOLCHAIN_GNU_EABI
  Building NuttX...
arm-none-eabi-ld: /github/workspace/sources/nuttx/nuttx section `.data' will not fit in region `flash'
arm-none-eabi-ld: region `flash' overflowed by 280 bytes
make[1]: *** [Makefile:174: nuttx] Error 1
make: *** [tools/Unix.mk:516: nuttx] Error 2
```

**Comment by pkarashchenko** [comment]
```
====================================================================================
Configuration/Tool: ea3131/pgnsh,CONFIG_ARM_TOOLCHAIN_GNU_EABI
------------------------------------------------------------------------------------
  Cleaning...
  Configuring...
  Disabling CONFIG_ARM_TOOLCHAIN_GNU_EABI
  Enabling CONFIG_ARM_TOOLCHAIN_GNU_EABI
  Building NuttX...
arm-none-eabi-ld: /github/workspace/sources/nuttx/nuttx section `.locked' will not fit in region `locked'
arm-none-eabi-ld: section .paged LMA [0000000011035000,0000000011042db7] overlaps section .locked LMA [0000000011029080,00000000110350fb]
arm-none-eabi-ld: section .ARM.exidx LMA [00000000110350fc,0000000011035103] overlaps section .paged LMA [0000000011035000,0000000011042db7]
arm-none-eabi-ld: region `locked' overflowed by 524 bytes
make[1]: *** [Makefile:174: nuttx] Error 1
make: *** [tools/Unix.mk:506: nuttx] Error 2
make: Target 'all' not remade because of errors.
```

**Comment by xiaoxiang781216** [comment]
@masayuki2009 @pkarashchenko this patchset try to extract the common operation from arch to sched which:

1. Avoid the same logic duplicate in each arch
2. Improve 10% sched performance

After the change, only when the current task need switch out, scheduler will call into arch code.

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by jerpelea — APPROVED** [review]

**Comment by pkarashchenko** [comment]
Please let me take a look. I will try to book some time slot in my calendar till Saturday

File: arch/xtensa/src/common/xtensa_blocktask.c [hunk 1]
```diff
@@ -45,108 +45,69 @@
...
 
-          /* Switch context to the context of the task at the head of the
-           * ready to run list.
-           */
+      /* Switch context to the context of the task at the head of the
+       * ready to run list.
+       */
 
-          nxsched_resume_scheduler(nexttcb);
+      nxsched_resume_scheduler(nexttcb);
```
Comment (pkarashchenko) [inline arch/xtensa/src/common/xtensa_blocktask.c:101]: Do we need
```
#ifdef CONFIG_ARCH_ADDRENV
              /* Make sure that the address environment for the previously
               * running task is closed down gracefully (data caches dump,
               * MMU flushed) and set up the address environment for the new
               * thread at the head of the ready-to-run list.
               */

              group_addrenv(nexttcb);
#endif
```
before calling `nxsched_resume_scheduler`?

**Comment by pkarashchenko** [comment]
@xiaoxiang781216 I'm comparing platform implementations of `up_reprioritize_rtr` and observe many differences. Are you sure that new generalised code is fitting all the supported archs?

**Comment by pkarashchenko** [comment]
Here is one example:
Screenshot 2022-11-18 at 00 39 34

**Comment by zyfeier** [comment]
> Here is one example: Screenshot 2022-11-18 at 00 39 34

@pkarashchenko up_reprioritize_rtr function was replaced by nxsched_reprioritize_rtr and up_block_task.  nxsched_reprioritize_rtr function is sched related process, and up_block_task is arch special process.

image

**Comment by masayuki2009** [comment]
@zyfeier @xiaoxiang781216 
Let me try to test some SMP use cases.

File: arch/xtensa/src/common/xtensa_blocktask.c (hunk 1 above)
Comment (xiaoxiang781216) [reply arch/xtensa/src/common/xtensa_blocktask.c:101]: I guess xtensa doesn't support kernel mode yet. And the origin code doens't invoke group_addrenv, it's better to not add group_addrenv in this patch.

File: arch/arm64/src/common/arm64_unblocktask.c [hunk 2]
```diff
@@ -43,102 +43,70 @@
...
-           * thread at the head of the ready-to-run list.
-           */
+      /* Make sure that the address environment for the previously
+       * running task is closed down gracefully (data caches dump,
+       * MMU flushed) and set up the address environment for the new
+       * thread at the head of the ready-to-run list.
+       */
 
-          group_addrenv(nexttcb);
+      group_addrenv(tcb);
```
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm64/src/common/arm64_unblocktask.c:94]: arm do this in the common place, e.g.:
https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/armv7-a/arm_doirq.c#L80-L90

[the remark this answers is not in the document]
Comment (pkarashchenko) [reply arch/arm64/src/common/arm64_unblocktask.c:94]: But ARM64 also have it at the exactly same place in `arm64_doirq`. So why calling in unblocktask is also needed?

**Review by pkarashchenko — COMMENTED** [review]
I do not have anything against this change. Only need to clarify `group_addrenv()` usage in general, but if overall behaviour is kept as before and change is well tested then `group_addrenv()` can be clarified separately from this PR

File: arch/arm64/src/common/arm64_unblocktask.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (xiaoxiang781216) [reply arch/arm64/src/common/arm64_unblocktask.c:94]: Sorry, I point to the wrong place. Please compare arm64_syscall_switch and arm_syscall for armv7-a, you can find that group_addrenv miss in arm64_syscall_switch.

**Comment by xiaoxiang781216** [comment]
> I do not have anything against this change. Only need to clarify `group_addrenv()` usage in general, but if overall behaviour is kept as before and change is well tested then `group_addrenv()` can be clarified separately from this PR

arm64 is fixed here: https://github.com/apache/incubator-nuttx/pull/7623. Other place need do group_addrenv in up_xxx  since the rest do the context switch directly without syscall.

**Comment by xiaoxiang781216** [comment]
> I do not have anything against this change. Only need to clarify `group_addrenv()` usage in general, but if overall behaviour is kept as before and change is well tested then `group_addrenv()` can be clarified separately from this PR

Ok, let's wait @masayuki2009 test result.

**Comment by xiaoxiang781216** [comment]
@masayuki2009 do you have the result?

**Comment by masayuki2009** [comment]
@xiaoxiang781216 
It seems working but I'd like to try the latest PR again.
So please fix the conflicts.

**Comment by zyfeier** [comment]
> @xiaoxiang781216 It seems working but I'd like to try the latest PR again. So please fix the conflicts.

@masayuki2009 conflicts was fixed,  please check again.

**Comment by masayuki2009** [comment]
@zyfeier @xiaoxiang781216 
I've just confirmed that all the tests have passed.
Let me merge this PR.

**Pull request merged**